File: genesis/src/main/java/net/spartanb312/genesis/java/builder/MethodBuilder.java
package net.spartanb312.genesis.java.builder;

import net.spartanb312.genesis.java.Utils;
import org.objectweb.asm.Label;
import org.objectweb.asm.tree.AnnotationNode;
import org.objectweb.asm.tree.MethodNode;

public abstract class MethodBuilder extends InsnListBuilder {

    public final MethodNode methodNode;

    public MethodBuilder(
            int access,
            String name,
            String desc,
            String signature,
            String[] exceptions
    ) {
        methodNode = new MethodNode(access, name, desc, signature, exceptions);
    }

    public MethodBuilder(
            int access,
            String name,
            String desc,
            String signature
    ) {
        methodNode = new MethodNode(access, name, desc, signature, null);
    }

    public MethodBuilder(
            int access,
            String name,
            String desc
    ) {
        methodNode = new MethodNode(access, name, desc, null, null);
    }

    public MethodNode toMethodNode() {
        methodNode.instructions = toInsnList();
        return methodNode;
    }

    public AnnotationNode addAnnotation(AnnotationNode annotation) {
        methodNode.visibleAnnotations = Utils.add(methodNode.visibleAnnotations, annotation);
        return annotation;
    }

    public AnnotationNode addAnnotation(AnnotationNode annotation, boolean visible) {
        if (visible) methodNode.visibleAnnotations = Utils.add(methodNode.visibleAnnotations, annotation);
        else methodNode.invisibleAnnotations = Utils.add(methodNode.invisibleAnnotations, annotation);
        return annotation;
    }

    public AnnotationNode addAnnotation(AnnotationBuilder builder) {
        AnnotationNode annotation = builder.toAnnotationNode();
        methodNode.visibleAnnotations = Utils.add(methodNode.visibleAnnotations, annotation);
        return annotation;
    }

    public AnnotationNode addAnnotation(AnnotationBuilder builder, boolean visible) {
        AnnotationNode annotation = builder.toAnnotationNode();
        if (visible) methodNode.visibleAnnotations = Utils.add(methodNode.visibleAnnotations, annotation);
        else methodNode.invisibleAnnotations = Utils.add(methodNode.invisibleAnnotations, annotation);
        return annotation;
    }

    public void MAXS(int maxStack, int maxLocals) {
        methodNode.visitMaxs(maxStack, maxLocals);
    }

    public void TRYCATCH(Label start, Label end, Label handler, String type) {
        methodNode.visitTryCatchBlock(start, end, handler, type);
    }

    public void LOCALVAR(String name, String descriptor, String signature, Label start, Label end, int index) {
        methodNode.visitLocalVariable(name, descriptor, signature, start, end, index);
    }

}

File: genesis/src/main/kotlin/net/spartanb312/genesis/MethodBuilder.kt
package net.spartanb312.genesis

import net.spartanb312.genesis.extensions.*
import org.objectweb.asm.Label
import org.objectweb.asm.tree.AnnotationNode
import org.objectweb.asm.tree.InsnList
import org.objectweb.asm.tree.MethodNode

@JvmInline
value class MethodBuilder(val methodNode: MethodNode) {

    operator fun InsnList.unaryPlus() = apply { methodNode.instructions.add(this) }
    operator fun AnnotationNode.unaryPlus() = apply { methodNode.visibleAnnotations.add(this) }
    operator fun InvisibleAnnotationNode.unaryPlus() = apply { methodNode.invisibleAnnotations.add(this) }

    @BuilderDSL
    fun MAXS(maxStack: Int, maxLocals: Int) = methodNode.visitMaxs(maxStack, maxLocals)

    @BuilderDSL
    fun INSTRUCTIONS(builder: InsnListBuilder.() -> Unit) {
        methodNode.instructions = instructions(builder)
    }

    @BuilderDSL
    fun TRYCATCH(start: Label, end: Label, handler: Label, type: String) {
        methodNode.visitTryCatchBlock(start, end, handler, type)
    }

}

@NodeDSL
inline fun MethodNode.modify(block: MethodBuilder.() -> Unit): MethodNode {
    MethodBuilder(this).block()
    return this
}

@NodeDSL
fun clinit(block: (MethodBuilder.() -> Unit)? = null): MethodNode = method(
    STATIC,
    "<clinit>",
    "()V", null,
    null,
).apply { if (block != null) modify(block) }

@NodeDSL
fun method(
    access: Modifiers,
    name: String,
    desc: String,
    signature: String? = null,
    exceptions: Array<String>? = null
): MethodNode = MethodNode(access.modifier, name, desc, signature, exceptions)

@NodeDSL
fun method(
    access: Int,
    name: String,
    desc: String,
    signature: String? = null,
    exceptions: Array<String>? = null
): MethodNode = MethodNode(access, name, desc, signature, exceptions)

@NodeDSL
inline fun method(
    access: Modifiers,
    name: String,
    desc: String,
    signature: String? = null,
    exceptions: Array<String>? = null,
    block: MethodBuilder.() -> Unit
): MethodNode = MethodNode(access.modifier, name, desc, signature, exceptions).modify(block)

@NodeDSL
inline fun method(
    access: Int,
    name: String,
    desc: String,
    signature: String? = null,
    exceptions: Array<String>? = null,
    block: MethodBuilder.() -> Unit
): MethodNode = MethodNode(access, name, desc, signature, exceptions).modify(block)